### PR TITLE
C Extension API 

### PR DESCRIPTION
**Pull request by samansmink**
This PR makes the first step towards proper support for https://github.com/duckdb/duckdb-rs/issues/370.
In the process it also bumps duckdb-rs to duckdb v1.1.1

With this PR I we introduce the experimental `duckdb_entrypoint_c_api` proc macro that will basically handle everything for you as can be seen in this snippet from `crates/duckdb/examples/hello-ext-capi`:

```rust
#[duckdb_entrypoint_c_api(ext_name = "rusty_quack", min_duckdb_version = "v0.0.1")]
pub unsafe fn ExtensionEntrypoint(con: Connection) -> Result<(), Box<dyn Error>> {
    con.register_table_function::<HelloVTab>("hello").expect("Failed to register hello table function");
    Ok(())
}
```

Note that this relies on the new extension C api added in https://github.com/duckdb/duckdb/pull/12682.

The main idea is that these binaries are (going to be) forwards compatible with new duckdb releases. This compatibility can be configured using the `min_duckdb_version` parament of the proc macro. For now, this version is set to a separate C extension API version (which is v0.0.1) in current duckdb stable. However in the future, this version will likely be simply the duckdb version. This means that loadable extensions currently produced with the `duckdb_entrypoint_c_api` proc are **not** stable right now. This stability *is* expected to be possible in DuckDB v1.2.0 in a few months.

Finally, the extension binaries in this PR are not ready to be loaded by DuckDB directly, they need to have their extension metadata appended. I will add this script to the https://github.com/duckdb/extension-ci-tools repo and create a rust extension template that uses the script to perform the final transform step for rust shared libary to loadable duckdb extension

### follow ups
- testing: the upgrade script running the bindgen performs no tests for the pre-generated bindings currently 
- create the demo extension template

### CODE REVIEW

**Comment by 0xcaff** [comment]
nice nice! I'll try to get this reviewed today evening or early tomorrow

File: crates/libduckdb-sys/upgrade.sh
Comment (0xcaff) [inline crates/libduckdb-sys/upgrade.sh:16]: we can remove the todo now

File: crates/libduckdb-sys/upgrade.sh
Comment (0xcaff) [inline crates/libduckdb-sys/upgrade.sh:32]: 🙈

Comment (0xcaff) [reply crates/libduckdb-sys/upgrade.sh:32]: iirc some of the tests are bindgen struct layout validations, not sure if these were failing or not

Comment (samansmink) [reply crates/libduckdb-sys/upgrade.sh:32]: Ok I will take a look, I think the problem here was that when building with the loadable_extension variant of the bindings, the tests have no way to create a duckdb instance making it basically impossible to run most of the tests

File: crates/duckdb-loadable-macros/src/lib.rs
Comment (0xcaff) [inline crates/duckdb-loadable-macros/src/lib.rs:65]: it will?

File: crates/duckdb/examples/hello-ext-capi/main.rs
Comment (0xcaff) [inline crates/duckdb/examples/hello-ext-capi/main.rs:25]: the default value of this pointer is not null, its some undefined value because the memory is allocated in C and passed across the ffi boundary. this branch will probably never be hit. seems like this free trait is implemented everywhere incorrectly in this project though so 🤷  just a heads up

Comment (0xcaff) [reply crates/duckdb/examples/hello-ext-capi/main.rs:25]: also your code doesn't ever not initialize name

File: crates/libduckdb-sys/build.rs
Comment (0xcaff) [inline crates/libduckdb-sys/build.rs:104]: nice can't miss this one

**Comment by samansmink** [comment]
Have a sample extension template here https://github.com/samansmink/extension-template-rs.

I think once we have this PR merged, I can:
- create duckdb/extension-template-rs
- create a new reusable workflow similar to `https://github.com/duckdb/extension-ci-tools/blob/main/.github/workflows/_extension_distribution.yml` but with just latest OS images everywhere since toolchain conpatibility no longer matters.
- Move makefile complexity out of the `extension-template-rs` into `extension-ci-tools` similar to how the c++ extension template does things.
- Add things like a rename script, extension uploading script, etc. 

Then later I need to think about:
- How to setup a testing framework for `extension-template-rs`? 
- Adding this to the community extensions so that people can distribute their Rust-only based extensions.

File: crates/libduckdb-sys/build.rs
Comment (0xcaff) [inline crates/libduckdb-sys/build.rs:304]: this looks suspicious. noticed i had it in my vendored version but not sure its correct as i don't think i ever exercise the codepath

File: crates/libduckdb-sys/build.rs
Comment (0xcaff) [inline crates/libduckdb-sys/build.rs:388]: might want to get rid of a few of these random commented out lines of code

File: crates/duckdb-loadable-macros/src/lib.rs
Comment (0xcaff) [inline crates/duckdb-loadable-macros/src/lib.rs:61]: access needs to be a `*const ffi::duckdb_extension_access` instead of a `ffi::duckdb_extension_access`

this somehow works on macos (i think due to the way arguments are passed in ARM on the stack sometimes) but breaks on x86_64 platforms in my testing. painful to debug this one

Comment (0xcaff) [reply crates/duckdb-loadable-macros/src/lib.rs:61]: for reference found it here https://github.com/0xcaff/duckdb_protobuf/pull/17/commits/7467ea6a31050385aabdb1108526dcaad5b9cf99

**Review by 0xcaff — CHANGES_REQUESTED** [review]
one blocking correctness issue (the pointer in duckdb_entrypoint_c_api's generated output is missing)

File: crates/duckdb-loadable-macros/src/lib.rs
Comment (0xcaff) [reply crates/duckdb-loadable-macros/src/lib.rs:61]: i think this is why the windows check is failing currently

**Comment by 0xcaff** [comment]
regarding tests for extensions, a contributor contributed tests to my project https://github.com/0xcaff/duckdb_protobuf/blob/master/packages/duckdb_protobuf/tests/it/main.rs might be useful as a scaffold though there are some rough edges to work out still

File: Makefile
Comment (samansmink) [reply Makefile:20]: We could do that, but given that we might want to do extension templates for other languages in the future as well, I thought it senses to just have python script to do this instead of a language specific script.

If you feel having a crate for just the extension sigining script makes sense I have no objections for adding them

File: crates/duckdb-loadable-macros/src/lib.rs
Comment (samansmink) [reply crates/duckdb-loadable-macros/src/lib.rs:65]: Extensions should not throw or exit the program on error, but instead use the method in the extension access struct to set an error, which duckdb will then throw as a C++ exception https://github.com/duckdb/duckdb/blob/2d0842be932e59a887c5decc79ba9d50842be7f7/src/main/extension/extension_load.cpp#L537

Comment (samansmink) [reply crates/duckdb-loadable-macros/src/lib.rs:65]: But now that I think about it I think about it, this is indeed incorrect at the moment:

I should just remove the error potentially returned from the `duckdb_rs_extension_api_init` function:
```rust
if #p_api.is_null() {
    return Err("DuckDB passed a nullpointer while trying to initialize the extension");
}
```

This case does indeed require no extra handling: DuckDB will only return a nullpointer if there is some error. But other parts of the initialization rust side might fail, which requires setting the error using the `extension_access` struct which i'm not doing here yet

File: crates/duckdb/examples/hello-ext-capi/main.rs
Comment (samansmink) [reply crates/duckdb/examples/hello-ext-capi/main.rs:25]: I just blindly copied `examples/hello-ext` I will leave it to a future pr to clean this up!

**Pull request merged**
